src/precomputed_quantities.jl

```diff
@@ -3,49 +3,340 @@
+        e_intⱼ = e_totⱼ - Kⱼ - Φ
+    end
+    if !(moisture_model isa DryModel)
+        q_totⱼ = divide_by_ρa(sgsⱼ.ρaq_tot, sgsⱼ.ρa, Yc.ρq_tot, Yc.ρ, a_min)
+        if moisture_model isa NonEquilMoistModel
+            q_liqⱼ = divide_by_ρa(sgsⱼ.ρaq_liq, sgsⱼ.ρa, Yc.ρq_liq, Yc.ρ, a_min)
+            q_iceⱼ = divide_by_ρa(sgsⱼ.ρaq_ice, sgsⱼ.ρa, Yc.ρq_ice, Yc.ρ, a_min)
+            q_ptⱼ = TD.PhasePartition(q_totⱼ, q_liqⱼ, q_iceⱼ)
+        end
+    end
+    if energy_form isa PotentialTemperature
+        if moisture_model isa DryModel
+            TD.PhaseDry_pθ(thermo_params, p, θⱼ)
+        elseif moisture_model isa EquilMoistModel
+            TD.PhaseEquil_pθq(thermo_params, p, θⱼ, q_totⱼ)
+        elseif moisture_model isa NonEquilMoistModel
+            TD.PhaseNonEquil_pθq(thermo_params, p, θⱼ, q_ptⱼ)
+        end
+    elseif energy_form isa TotalEnergy
+        if moisture_model isa DryModel
+            TD.PhaseDry_pe(thermo_params, p, e_intⱼ)
+        elseif moisture_model isa EquilMoistModel
+            TD.PhaseEquil_peq(thermo_params, p, e_intⱼ, q_totⱼ)
+        elseif moisture_model isa NonEquilMoistModel
+            TD.PhaseNonEquil_peq(thermo_params, p, e_intⱼ, q_ptⱼ)
+        end
+    end
+end
+
+
+"""
+    set_precomputed_quantities!(Y, p, t)
+
+Uses the prognostic state `Y` to set the values of `ᶜu`, `ᶠu³`, `ᶜK`, `ᶜts`, and
+`ᶜp`. If `turbconv_model` is `EDMFX`, also sets the values of `ᶜuⱼs`, `ᶠu³ⱼs`,
+`ᶜKⱼs`, `ᶜtsⱼs`, `ᶜρa₀`, `ᶠw₀`, `ᶜu₀`, `ᶠu³₀`, `ᶜK₀`, and `ᶜts₀`.
+
+This function also applies a "filter" to `Y` in order to ensure that `ᶠu³` is 0
+at the surface (i.e., to enforce the impenetrable boundary condition). If
+`turbconv_model` is `EDMFX`, the filter also ensures that the `ᶠu³ⱼs` and `ᶠu³₀`
+are 0 at the surface. In the future, we may want to move this filtering
+elsewhere, but doing it here ensures that it occurs whenever the precomputed
+quantities are updated.
+
+Note: If you need to update the value of any of the aforementioned quantities,
+please call this function instead of recomputing the value yourself. Otherwise,
+it will be difficult to ensure that the duplicated computations are consistent.
+"""
+function set_precomputed_quantities!(Y, p, t)
+    (; energy_form, moisture_model, turbconv_model) = p.atmos
+    thermo_params = CAP.thermodynamics_params(p.params)
+    (; ᶜΦ) = p
+    (; a_min) = turbconv_model
+    ᶠinterp = Operators.InterpolateC2F(
+        bottom = Operators.Extrapolate(),
+        top = Operators.Extrapolate(),
+    )
+
+    tsⱼs(Yc, Kⱼs, p, Φ) =
+        map(Yc.sgsⱼs, Kⱼs) do sgsⱼ, Kⱼ
+            tsⱼ(
+                sgsⱼ,
+                Yc,
+                Kⱼ,
+                p,
+                Φ,
+                a_min,
+                thermo_params,
+                energy_form,
+                moisture_model,
+            )
+        end
+
+    w₀(ρaⱼs, wⱼs, ρ, w) =
+        divide_by_ρa(ρ * w - sum(ρaⱼs .* wⱼs), ρ - sum(ρaⱼs), ρ * w, ρ, a_min)
```

Comment:
We'll probably need `ρ0 = ρ - sum(ρaⱼs)` in other places too (e.g., in diffusion terms in GS equations), so it may make sense to cache it before this computation.